### DISCUSSION TITLE
Issue connecting circuits together

### DISCUSSION THREAD

**Discussion by snbamford** [netbox-community/netbox#13858] [Q&A]
We are often in a situation where we want to connect circuits directly together. For example, a core circuit running directly from one location to another, with one end of the core circuit connected to the rear port on a patch-patch or directly to a device interface, the other end connected to a third-party port that we represent as a circuit with only one termination.

Often when we try to connect the two circuits directly together, we get an error message that says:
duplicate key value violates unique constraint "circuits_circuittermination_connected_endpoint_id_key"
DETAIL:  Key (connected_endpoint_id)=(13165) already exists.

Both circuit end-points are available when we get this error. We've been able to work around the issue most times by deleting and re-creating one or both circuits exactly the same as it was before, but there are occasions where this doesn't work.

Has anyone else encountered this issue? I haven't had any luck finding any reference to the error message online when searching in the past.

Kind regards,
Scott
![image](https://github.com/netbox-community/netbox/assets/2368549/cd7f615c-d3f1-47a0-82bf-9869eb8597a8)

**Comment by candlerb** [comment]
There have been a lot of changes to the cabling models since Netbox v2.9 - for example [v3.3](https://github.com/netbox-community/netbox/releases/v3.3.0) allows one cable to have multiple terminations at each end - and lots of bug fixes.

Whilst it's quite likely that you've found a bug in v2.9.11 (to classify it as such you'd have to provide the exact steps to reproduce it), the Netbox project doesn't create maintenance fixes for old versions. All fixes are made on the head only, and once v2.10.0 is released, no more v2.9.x releases are made.

Therefore I would suggest you test to see whether a newer version works better for you:
* build a test instance with v2.11.12
* [export your postgres data](https://docs.netbox.dev/en/stable/administration/replicating-netbox/) to this test instance and re-run 'upgrade.sh' to apply schema updates
* test it
* upgrade this test instance to v3.6.2
* test it again
* if this solves your problem, upgrade your production instance

(Note that you must go via v2.11.x; versions v3.0.0 and later can only upgrade from v2.11.0 onwards)

The real pain in this upgrade will be if you have any existing integrations (e.g. REST API integration, custom scripts, reports, webhooks); they will all need changing to match changes in the APIs, which have no backwards-compatibility guarantees. This means you'll have to read the release notes for the .0 point releases of all intervening versions: [v2.10.0](https://github.com/netbox-community/netbox/releases/v2.10.0), [v2.11.0](https://github.com/netbox-community/netbox/releases/v2.11.0), [v3.0.0](https://github.com/netbox-community/netbox/releases/v3.0.0) etc and look for "breaking changes".
